storage/src/state/operator.rs

```diff
@@ -0,0 +1,187 @@
+        self.shares.add_shares(block_height, address, num_shares)
+    }
+
+    /// Removes the shares for a given block height in storage.
+    pub fn remove_shares(&self, block_height: u32) -> Result<()> {
+        self.shares.remove_shares(block_height)
+    }
+
+    /// Returns `true` if the given commitment exists in storage.
+    pub fn contains_coinbase_record(&self, commitment: &N::Commitment) -> Result<bool> {
+        self.shares.contains_record(commitment)
+    }
+
+    /// Returns all coinbase records in storage.
+    pub fn to_coinbase_records(&self) -> Vec<(u32, Record<N>)> {
+        self.shares.to_records()
+    }
+
+    /// Returns the coinbase record for a given commitment.
+    pub fn get_coinbase_record(&self, commitment: &N::Commitment) -> Result<(u32, Record<N>)> {
+        self.shares.get_record(commitment)
+    }
+
+    /// Adds the given coinbase record to storage.
+    pub fn add_coinbase_record(&self, block_height: u32, record: Record<N>) -> Result<()> {
+        self.shares.add_record(block_height, record)
+    }
+
+    /// Removes the given record from storage.
+    pub fn remove_coinbase_record(&self, commitment: &N::Commitment) -> Result<()> {
+        self.shares.remove_record(commitment)
+    }
+}
+
+#[derive(Clone, Debug)]
+#[allow(clippy::type_complexity)]
+struct SharesState<N: Network> {
+    /// The miner shares for each block height.
+    shares: DataMap<u32, HashMap<Address<N>, u64>>,
+    /// The coinbase records earned by the operator.
+    records: DataMap<N::Commitment, (u32, Record<N>)>,
+}
+
+impl<N: Network> SharesState<N> {
+    /// Initializes a new instance of `SharesState`.
+    fn open<S: Storage>(storage: S) -> Result<Self> {
+        Ok(Self {
+            shares: storage.open_map("shares")?,
+            records: storage.open_map("records")?,
+        })
+    }
+
+    /// Returns all shares in storage.
+    fn to_shares(&self) -> Vec<(u32, HashMap<Address<N>, u64>)> {
+        self.shares.iter().collect()
+    }
+
+    /// Returns the shares for a given block height.
+    fn get_shares(&self, block_height: u32) -> Result<HashMap<Address<N>, u64>> {
+        match self.shares.get(&block_height)? {
+            Some(shares) => Ok(shares),
+            None => return Err(anyhow!("Block height {} does not have any shares in storage", block_height)),
+        }
+    }
+
+    /// Adds the given number of shares to the block height and address in storage.
+    fn add_shares(&self, block_height: u32, address: &Address<N>, num_shares: u64) -> Result<()> {
+        if let Some(current_shares) = self.shares.get(&block_height)? {
+            let mut new_shares = current_shares.clone();
+
+            // Add the num shares for the address.
+            let address_entry = new_shares.entry(*address).or_insert(0);
+            *address_entry = address_entry.saturating_add(num_shares);
+
+            // Insert the shares for the address.
+            self.shares.insert(&block_height, &new_shares)?;
```

Comment:
why not just `get_mut(&block_height)` and update that value instead of doing a lookup _and_ an `insert` at the end?

Reply:
Again, doesn't seem to apply to the rocksdb `DataMap`.

Reply:
ah, fair, I mixed it up with a `HashMap` :+1: